src/krb5_plugin/passkey/passkey_kdcpreauth.c

```diff
@@ -0,0 +1,429 @@
+                     krb5_kdc_req *request,
+                     krb5_kdcpreauth_callbacks cb,
+                     krb5_kdcpreauth_rock rock,
+                     krb5_kdcpreauth_moddata moddata,
+                     krb5_preauthtype pa_type,
+                     krb5_kdcpreauth_edata_respond_fn respond,
+                     void *arg)
+{
+    struct sss_passkeykdc_config *config = NULL;
+    struct sss_radiuskdc_state *state;
+    krb5_keyblock *armor_key;
+    char *configstr = NULL;
+    krb5_error_code ret;
+
+    state = (struct sss_radiuskdc_state *)moddata;
+
+    ret = sss_radiuskdc_enabled(SSSD_PASSKEY_CONFIG, kctx, cb, rock, &configstr);
+    if (ret != 0) {
+        goto done;
+    }
+
+    armor_key = cb->fast_armor(kctx, rock);
+    if (armor_key == NULL) {
+        ret = ENOENT;
+        goto done;
+    }
+
+    ret = sss_passkeykdc_config_init(state, kctx, cb->client_name(kctx, rock),
+                                     configstr, &config);
+    if (ret != 0) {
+        goto done;
+    }
+
+    ret = sss_passkeykdc_challenge_send(kctx, cb, rock, respond, arg,
+                                        config->radius);
+
+done:
+    if (ret != 0) {
+        respond(arg, ret, NULL);
+    }
+
+    cb->free_string(kctx, rock, configstr);
+    sss_passkeykdc_config_free(config);
+}
+
+static void
+sss_passkeykdc_verify(krb5_context kctx,
+                      krb5_data *req_pkt,
+                      krb5_kdc_req *request,
+                      krb5_enc_tkt_part *enc_tkt_reply,
+                      krb5_pa_data *pa,
+                      krb5_kdcpreauth_callbacks cb,
+                      krb5_kdcpreauth_rock rock,
+                      krb5_kdcpreauth_moddata moddata,
+                      krb5_kdcpreauth_verify_respond_fn respond,
+                      void *arg)
+{
+    struct sss_radiuskdc_state *state;
+    struct sss_passkeykdc_config *config = NULL;
+    struct sss_passkey_message *message = NULL;
+    char *configstr = NULL;
+    krb5_error_code ret;
+    krb5_data rstate;
+
+    state = (struct sss_radiuskdc_state *)moddata;
+
+    ret = sss_radiuskdc_enabled(SSSD_PASSKEY_CONFIG, kctx, cb, rock, &configstr);
+    if (ret != 0) {
+        goto done;
+    }
+
+    ret = sss_passkeykdc_config_init(state, kctx, cb->client_name(kctx, rock),
+                                     configstr, &config);
+    if (ret != 0) {
+        goto done;
+    }
+
+    ret = sss_radiuskdc_get_cookie(kctx, cb, rock, SSSD_PASSKEY_PADATA,
+                                   &rstate);
+    if (ret != 0) {
+        goto done;
+    }
+
+    if (pa->pa_type != SSSD_PASSKEY_PADATA || pa->length != 0) {
+        ret = KRB5_PREAUTH_BAD_TYPE;
+        goto done;
+    }
+
+    message = sss_passkey_message_decode_padata(pa);
+    if (message == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    if (message->phase != SSS_PASSKEY_PHASE_REPLY
+        || strlen(message->state) != rstate.length
```

Comment:
Hi,

`rstate.length` includes the trailing `\0` byte, so the check should be `strlen(message->state) + 1 != rstate.length`.

bye,
Sumit